DOC: FAQ for cov estimation

dengemann opened the pull request:
First pass ....
@christianbrodbeck @Eric89GXL @agramfort

Eric89GXL commented:
Relevant references we should keep in mind and maybe close once this is done: #1728 #1675 #1637 #1495

Eric89GXL commented:
Looks reasonable to me. Can you add "Closes XXX" to your main description as you eliminate items from the above issues?

Eric89GXL commented:
Looks like a good start. It should probably at least address all of @christianbrodbeck's questions from #1637 and #1728 -> #1675, too. Can we have headings and sub-headings that get rendered appropriately? This could mostly contain things under the `Covariance estimation` heading, for example, since he had many questions about that. And later we could add other headings that people bring up, e.g. from

teonlamont commented on doc/source/faq.rst:
```diff
@@ -0,0 +1,76 @@
+These methods for evaluation can be used to assess model violations. Additional
+introductory materials cen be found [here](https://speakerdeck.com/dengemann/eeg-sensor-covariance-using-cross-validation).
+
+For expert use cases or debugging the alternative estimators can also be compareded::
```
compareded --> compared

dengemann commented:
closes #1934 #1728 #1675 #1637 #1495.

Eric89GXL commented:
hah! that's the greatest `GitHub` comment I think I've ever seen. Have you really made that much progress? If so, @agramfort owes you a beer and you can have one of the beers he owes me, too.

dengemann commented:
@Eric89GXL I blindly followed your suggestion ... let me check now :)

Eric89GXL commented:
I meant to add the "closes" only once you had actually added the content to the PR to justify closing the issue... I don't think your first commit quite got us that far :)

dengemann commented:
@Eric89GXL updated the list. I think we're ok for now. Relevance - complexity tradeoff

Eric89GXL commented:
I'm starting to think we won't get to #1495 before 0.9 release, unless someone steps forward to dedicate the necessary hours to it. Should we change that tag to 0.10? Same with #1934 perhaps?

Eric89GXL commented:
...in any case, let me know if you're done here (I think you probably are?) and I'll review/merge.

dengemann commented:
> I'm starting to think we won't get to #1495 before 0.9 release, unless someone steps forward to dedicate the necessary hours to it. Should we change that tag to 0.10? Same with #1934 perhaps?

I updated the list and removed these two as I came to the same conclusions as you did!

Eric89GXL commented on doc/source/faq.rst:
```diff
@@ -0,0 +1,76 @@
+
+    >>> import mne
+    >>> epochs = mne.read_epochs(epochs_path)
+    >>> cov = mne.compute_covariance(epochs, tmax=0., method='auto')
```
Did you check to make sure this rendered okay? I think you might need a `::` in the preceding line for it to know this is a code literal. GitHub also seems to render it differently than the code literal below, which makes me think you do need `::`.

dengemann replied:
let me doc build. I also might want to add one sentence on the reasons for which we recommend regularization.

Eric89GXL replied:
...and did you end up checking this?

dengemann replied:
I did but apparently I've missed this one ...

Eric89GXL commented on doc/source/faq.rst:
```diff
@@ -0,0 +1,76 @@
+
+The MNE manual suggests regularizing the noise covariance matrix (see
+:ref:`CBBHEGAB`). In MNE-Python, this is done using the regularization methods
+described in [1, 2, 3, 4]. For this the 'auto' option can be used. With this
```
And is this the proper way to make references to the articles below? I thought you'd need some `_`'s in there.

dengemann replied:
Not sure, do we have an exmaple somewhere?

Eric89GXL replied:
Not sure if we do, but try this:

https://github.com/scipy/scipy/blob/master/scipy/signal/signaltools.py#L1989

Eric89GXL commented on doc/source/faq.rst:
```diff
@@ -0,0 +1,76 @@
+Inverse Solution
+================
+
+How should I regularize the covariance matrix
```
I think this should actually end with a question mark, I think most FAQs do that

Eric89GXL commented:
Thanks for taking care of this (along with @christianbrodbeck) -- I think it will be good to have it started so we can add to it incrementally.

agramfort commented:
+1 for merged avec @Eric89GXL 's comments have been addressed.

Eric89GXL commented:
FYI for others who read this later Google Translate says "avec" in French means "with" in English

agramfort commented:
oups... I am a bit tired... fixed...

dengemann commented:
@Eric89GXL @agramfort I'd like to remove the usage of regularize in our examples ...
Option 1) trust on the rank-based regularizer which we have inside the whitening workflow. 2) explicitly compute the cov in these examples 3) recompute the cov we ship with the dataset using the auto option.

Eric89GXL commented:
I am fine with the path of least resistance for now, which I guess would be 1).

dengemann commented:
> which I guess would be 1).

according to our latest research the sparse solvers might not be happy with this. But maybe it's not that bad as we have sufficient numners of trials.

dengemann commented:
@Eric89GXL bad news the gamma map is not happy without additional regularization

https://www.dropbox.com/s/mtlx9ube6m8iza5/Screenshot%202015-04-20%2004.35.22.png?dl=0

dengemann commented:
With shrunk results are a bit different than so far but still similar. And it's the more principled approach.

https://www.dropbox.com/s/w7ghgfic1zfzjpj/Screenshot%202015-04-20%2004.50.48.png?dl=0

dengemann commented:
Maybe we should just compute a regularized covariance based on 'auto' ... It would allow us not to blow up each example with cov code.

Eric89GXL commented:
Well I think we do want to allow people who have covariances calculated the old-fashioned way to still be able to use those if they wish, so maybe the examples that need regularization can still make use of the basic channel-type-based regularization?

Eric89GXL commented on doc/source/faq.rst:
```diff
@@ -0,0 +1,73 @@
+:ref:`CBBHEGAB`), especially if only few samples are available. Unfortunately
+it is not easy to tell the effective number of samples, hence, to chose the appropriate regularization.
+In MNE-Python, regularization is done using advanced regularization methods
+described in [1]. For this the 'auto' option can be used. With this
```
Don't forget to change this to `[1]_` a la scipy.

dengemann replied:
missed this one.

dengemann commented:
> so maybe the examples that need regularization can still make use of the basic channel-type-based regularization?

I think we wanted to deprecate the regularize function. We can keep it for manual interventions of couse. But I think it's not the right way to advertise manual tuning if we have a principled solution. So I'm rather -1 for exposing it.

Eric89GXL commented:
How long does it take to use `auto` on your computer? On some of mine it takes a while, which would take our examples in some cases from pretty quick to very slow.

dengemann commented:
> How long does it take to use auto on your computer?

A few seconds only. But adding a regularized cov should not be too expensive in terms of storage.

Eric89GXL commented:
@dengemann is this good to be merged from your end, or do you want to deal with the example updates as part of this PR?

dengemann commented:
@Eric89GXL fine with merging. But I'll have to open a new issue then.

Eric89GXL commented:
Let's do that in case we want to add other things to the FAQ. Plus we might need to discuss the best path forward with the cov stuff anyway.

Eric89GXL commented:
Thanks for getting us going on this!
